game-core/src/main/java/games/strategy/triplea/ui/SimpleUnitPanel.java

```diff
@@ -205,8 +230,14 @@ private void addUnits(
       icon.ifPresent(label::setIcon);
       MapUnitTooltipManager.setUnitTooltip(label, unitType, player, quantity);
     } else if (unit instanceof Resource) {
-      label.setIcon(uiContext.getResourceImageFactory().getIcon(unit, true));
+      label.setIcon(uiContext.getResourceImageFactory().getIcon(unit, useLargeIcons));
     }
     add(label);
   }
+
+  private void addEmptyLabelIfNeeded() {
+    if (emptyLabel != null && getComponentCount() == 0) {
```

Comment:
`emptyLabel` seems to be violating SRP. Specifically it serves two purposes:
1) Not-null means we should display it when there are no components
2) The value itself is the display component

So we are using both the display element as both the display element and the flag whether to show the display element. It's also a bit notable too that it is mutable with 'setEmptyLabel'.

Do you agree with this issue @asvitkine and have you had any thoughts about it?

Looking at the diff, as a suggestion, I think you can solve this by adding a constructor overload that accepts a new parameter, an enum "EmptyLabelDisplayMode". The existing constructor would invoke that new parameter

EG:
```
public SimpleUnitPanel(final UiContext uiContext, final boolean useSmallIconWrapLayout) {
   this(uiContext, useSmallIconWrapLayout, EmptyLabelDisplayMode.DO_NOT_DISPLAY);
}

public SimpleUnitPanel(final UiContext uiContext, final boolean useSmallIconWrapLayout, EmptyLabelDisplayMode emptyLabelDisplayMode) {
  :
  :
}
```

I also think you can then have 'SimpleUnitPanel' own the 'empty' label and not have it be injected. Then this add method would be:

```
private void addEmptyLabelIfNeeded() {
  if(emptyLabelDisplayMode == DISPLAY_ON_EMPTY_PURCHASE && getComponentCount() == 0) {
      add(new JLabel("(None)");
 }
}
```

Reply:
To be honest, I can't say I'm huge fan of your suggested implementation, but on the other hand I don't feel very strongly about it and so happy to change it to whatever fits the project style conventions.

But to expand, within a class it seems reasonable to use a member's null state in this way - it keeps the code simpler and shorter and not ambiguous (we could add a comment about what null means above the line).

I agree that the label being mutable can be a little weird - and makes the API more powerful than strictly needed. In fact, even passing it to the ctor would still allow the caller to retain a reference and modify it, if needed. If we wanted to restrict it, we could make a String param to the ctor to specify the label text instead, but it won't address your SRP concerns with nullability.

But as I said, I'm OK going with an enum approach too. However, since we're already adding as useSmallIconsWrapLayout param anyway, as a boolean (that could be an enum), one option would be to combine them them into just one enum that specifies the 'mode' or 'style' of the component. (Technically the new boolean already combines two things anyway - the icon size and the layout.) So having a high level enum specifying a number of behaviours seems reasonable given these are the only uses.

The enum could be something like:
DisplayMode {
  LARGE_ICONS_COLUMN
  SMALL_ICONS_WRAPPED_WITH_LABEL_WHEN_EMPTY
}

WDYT?

Reply:
I went ahead and tried that out. Let me know what you think.